src/dwarf/debug_info.rs

```diff
@@ -0,0 +1,1325 @@
+
+        let abbrev = unsafe { mem::transmute(&self.abbrevs[abbrev_idx as usize - 1]) };
+        self.abbrev = Some(abbrev);
+        if abbrev.has_children {
+            self.cur_depth += 1;
+        }
+
+        self.die_reading_done = false;
+        Some(DIE {
+            tag: abbrev.tag,
+            offset: saved_off as u64,
+            abbrev: Some(abbrev),
+            abbrev_attrs: abbrev.all_attrs(),
+            abbrev_attrs_idx: 0,
+            data: &self.data[self.off..],
+            dieiter: unsafe { mem::transmute(self) },
+            reading_offset: 0,
+            done: false,
+        })
+    }
+}
+
+/// An iterator of Units in the `.debug_info` section.
+///
+/// An iterator is built from the content of `.debug_info` section,
+/// which is a list of compile units.  A compile unit usually refers
+/// to a source file.  In the compile units, it is a forest of DIEs,
+/// which presents functions, variables and other debug information.
+pub struct UnitIter<'a> {
+    info_data: &'a [u8],
+    abbrev_data: &'a [u8],
+    pub off: usize,
+}
+
+impl<'a> UnitIter<'a> {
+    /// Build an iterator from the content of `.debug_info` & `.debug_abbrev`.
+    ///
+    /// # Arguments
+    ///
+    /// * `info_data` is the content of the `.debug_info` section.
+    /// * `abbrev_data` is the content of the `.debug_abbrev` section.
+    pub fn new(info_data: &'a [u8], abbrev_data: &'a [u8]) -> UnitIter<'a> {
+        UnitIter {
+            info_data,
+            abbrev_data,
+            off: 0,
+        }
+    }
+}
+
+impl<'a> Iterator for UnitIter<'a> {
+    type Item = (UnitHeader, DIEIter<'a>);
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let off = self.off;
+        let uh = parse_unit_header(&self.info_data[off..])?;
+        let hdr_sz = uh.header_size();
+        self.off += uh.unit_size();
+
+        match uh {
+            UnitHeader::CompileV4(ref cuh) => {
+                let dwarf_sz = if cuh.bits64 { 8 } else { 4 };
+                let addr_sz = cuh.address_size as usize;
+                let (abbrevs, _) =
+                    parse_cu_abbrevs(&self.abbrev_data[cuh.debug_abbrev_offset as usize..])?;
+                Some((
+                    uh,
+                    DIEIter {
+                        data: &self.info_data[off + hdr_sz..],
+                        dwarf_sz,
+                        addr_sz,
+                        off: 0,
+                        off_delta: hdr_sz,
+                        cur_depth: 0,
+                        abbrevs,
+                        abbrev: None,
+                        die_reading_done: true,
+                        done: false,
+                    },
+                ))
+            }
+            UnitHeader::CompileV5(ref _cuh) => {
+                todo!(); // BlazeSym supports only v4 so far.
+            }
+            _ => self.next(),
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    const TEST_INFO: [u8; 0x80a] = [
```

Comment:
this is practically undebuggable, how would someone make sense of this binary blob? Have you considered checking in a small ELF file with .debug_info and .debug_abbrev in it (which can be inspected with readelf and llvm-dwarfdump, if necessary) and using that from unit tests?

Reply:
Use objcopy to create an ELF including only DWARF sections.

Reply:
so did you generate ELF first and then dumped DWARF binary here? Or you hand-crafted DWARF just for this test?

Because if the former and you started with ELF, I'm saying let's check in those ELF files and use them in unit tests to get raw DWARF data programmatically.

The point is that ELF is a data container you can work with with generic tools like objcopy/readelf/llvm-dwarfdump. While hard-coded array in Rust code is not that and it makes it very hard to a) understand what's the contents of this binary data and b) change/debug it.